### PR TITLE
refactor(optimizer): column prune refactor

### LINKED ISSUE #2054: refactor: replace `bitset` with vector when performing column pruning
Currently, when performing column pruning in the optimizer, we use a `FixedBitset` to indicate which columns should be kept. However, this approach is unable to adjust the order of the output columns. This blocks the implementation of projection elimination and output index. It would be better if we unify them into a single `Vec`.

We should replace `&FixedBitset` with `&[usize]` in `col_prune`.

## Dependencies

#1922 depends on this issue.

### PR DESCRIPTION
**Pull request by Enter-tainer**
### Summarize your change
This pr mainly change `prune_col`'s type signature from `fn prune_col(&self, required_cols: &FixedBitSet) -> PlanRef` to `fn prune_col(&self, required_cols: &[usize]) -> PlanRef`. It also comes with refactors and readablity improvements on `LogicalAgg`, `LogicalFilter`, `LogicalProject`, `LogicalTopn`.

### How does this PR work?

This pr is the very first step of #1922. By switching from `FixedBitset` to `Vec`, we will be able to adjust the order of the columns when performing column pruning. For example, `required_cols = [3, 1]` means "keeping the 3rd column and the 1st column, and put the 3rd column after the 1st column". Note that reordering is currently not implemented in this pr, it will be added in the future possibly by inserting an extra projection node if needed. The correctness is not affected anyway because we currently do not use the reorder function.

After this pr, we can implement `output index` described in #1922.

### Describe any limitations of the current code 
There is still some cleaning work to do, and I'm still working on this.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)
close #2054, related: #1922

### CODE REVIEW

**Comment by st1page** [comment]
Is it ready for review?

File: src/frontend/src/expr/utils.rs [hunk 1]
```diff
@@ -372,10 +372,11 @@ impl CollectInputRef {
             input_bits: FixedBitSet::with_capacity(capacity),
         }
     }
+}
 
-    /// Returns the collected indexes by the `CollectInputRef`.
-    pub fn collect(self) -> FixedBitSet {
```
Comment (st1page) [inline src/frontend/src/expr/utils.rs:377]: in my private opinion `collect` is more readable, but both is ok

**Review by st1page — APPROVED** [review]
LTGM, good work! let's implement the columns reordering then.

File: src/frontend/src/expr/utils.rs (hunk 1 above)
Comment (Enter-tainer) [reply src/frontend/src/expr/utils.rs:377]: You are right, Into or From sometimes requires extra type hint.

**Pull request merged**